oeps/oep-38-Data-Modeling.rst

```diff
@@ -0,0 +1,438 @@
+      remove the data from the system using the SQL delete keyword.
+
+   -  Please note that GDPR may require the data be deleted. If a field
+         is determined to contain PII and falls under the realm of GDPR,
+         That data should be deleted from the system.
+
+Don’t trap the data
+
+-  Each piece of information should have its own column. Avoid storing
+      data in blob fields or as JSON in the database.
+
+-  Don’t store encoded (pickle, json, other) objects in the database! If
+      you need to run the python environment to decode the data, the
+      data is worthless.
+
+Store everything
+
+-  Storage is cheap!
+
+-  If you are unsure whether you should store something in the database,
+      add history, or fire an event, the answer is almost always yes
+
+-  Still not sure? The default answer is yes.
+
+Views should access models via methods on those models, instead of
+querying managers directly.
+
+-  For example, prefer creating something like
+      CourseEnrollment.is_enrolled(...) rather than having views check
+      CourseEnrollment.objects.filter(...).exists().
+
+-  This allows us to more easily make internal representation changes
+      like switching to using a “deleted” flag instead of deleting the
+      row.
+
+-  This also reduces the likelihood that people will query models in a
+      non-performant way (e.g. sorting by an unindexed field).
+
+Enforce logical constraints at the database layer.
+
+-  If your code expects a 1:1 relationship, make that a unique
+      constraint in the database instead of trying to just enforce it in
+      Python.
+
+-  Python will not save you from race conditions. Database constraints
+      will.
+
+-  For example, an enrollment should have a unique constraint on
+      (course_id, user_id), since a given user should only have one
+      enrollment per course.
+
+-  This is a smaller version of “don’t allow impossible states to be
+      represented in the database”.
+
+Developer Responsibility
+
+------------------------
+
+It is the responsibility of the developer to adhere to all of the
+standards in the Data Modeling Standards section of this document.
+
+Code Reviewer Responsibility
+
+----------------------------
+
+The code reviewer is responsible for ensuring the standards set forth in
+the Data Modeling Standards section of this document are met.
+
+Responsibility for Third-party Service Integrations
+
+---------------------------------------------------
+
+Not sure… Adhere to the same standards.
+
+Backward Compatibility
+
+======================
+
+Data models that are not within the standards of this document should be
+updated the next time a team needs to make changes to that model
+
+Some things are exempt from this process:
+
+-  Do not update primary key types without business/technical
+      justification, as there may be many downstream changes that will
+      need to change as well.
+
+Reference Implementation
```

Comment:
I think illustrative examples of good and bad models (either by linking or inlining some code) would make the OEP more impactful.